File: app/src/main/java/me/zhanghai/android/douya/ui/RatioImageView.java
/*
 * Copyright (c) 2015 Zhang Hai <Dreaming.in.Code.ZH@Gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.douya.ui;

import android.content.Context;
import android.os.Build;
import android.util.AttributeSet;

/**
 * An ImageView that measures with a ratio. Also sets scaleType to centerCrop.
 */
public class RatioImageView extends AdjustViewBoundsImageView {

    private float mRatio;

    public RatioImageView(Context context) {
        super(context);

        init();
    }

    public RatioImageView(Context context, AttributeSet attrs) {
        super(context, attrs);

        init();
    }

    public RatioImageView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);

        init();
    }

    private void init() {
        setScaleType(ScaleType.CENTER_CROP);
    }

    public float getRatio() {
        return mRatio;
    }

    public void setRatio(float ratio) {
        if (mRatio != ratio) {
            mRatio = ratio;
            requestLayout();
            invalidate();
        }
    }

    public void setRatio(float width, float height) {
        setRatio(width / height);
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {

        if (mRatio > 0) {
            if (MeasureSpec.getMode(heightMeasureSpec) == MeasureSpec.EXACTLY) {
                int height = MeasureSpec.getSize(heightMeasureSpec);
                int width = Math.round(mRatio * height);
                width = Math.max(width, getSuggestedMinimumWidth());
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
                    width = Math.min(width, getMaxWidth());
                }
                widthMeasureSpec = MeasureSpec.makeMeasureSpec(width, MeasureSpec.EXACTLY);
            } else {
                int width = MeasureSpec.getSize(widthMeasureSpec);
                int height = Math.round(width / mRatio);
                height = Math.max(height, getSuggestedMinimumHeight());
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
                    height = Math.min(height, getMaxHeight());
                }
                heightMeasureSpec = MeasureSpec.makeMeasureSpec(height, MeasureSpec.EXACTLY);
            }
        }

        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
    }
}

File: app/src/main/java/me/zhanghai/android/douya/ui/RatioImageView.kt
/*
 * Copyright (c) 2020 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.douya.ui

import android.content.Context
import android.util.AttributeSet
import kotlin.math.roundToInt

/**
 * An [android.widget.ImageView] that measures with a ratio. Also sets scaleType to centerCrop.
 */
class RatioImageView : AdjustViewBoundsImageView {
    constructor(context: Context) : super(context)

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs)

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(
        context, attrs, defStyleAttr
    )

    init {
        scaleType = ScaleType.CENTER_CROP
    }

    var ratio = 1f
        set(ratio) {
            if (field == ratio) {
                return
            }
            field = ratio
            requestLayout()
            invalidate()
        }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        var coercedWidthMeasureSpec = widthMeasureSpec
        var coercedHeightMeasureSpec = heightMeasureSpec
        if (ratio > 0) {
            if (MeasureSpec.getMode(coercedHeightMeasureSpec) == MeasureSpec.EXACTLY) {
                val height = MeasureSpec.getSize(coercedHeightMeasureSpec)
                val width = (ratio * height).roundToInt().coerceIn(suggestedMinimumWidth, maxWidth)
                coercedWidthMeasureSpec = MeasureSpec.makeMeasureSpec(width, MeasureSpec.EXACTLY)
            } else {
                val width = MeasureSpec.getSize(coercedWidthMeasureSpec)
                val height = (width / ratio).roundToInt().coerceIn(
                    suggestedMinimumHeight, maxHeight
                )
                coercedHeightMeasureSpec = MeasureSpec.makeMeasureSpec(height, MeasureSpec.EXACTLY)
            }
        }
        super.onMeasure(coercedWidthMeasureSpec, coercedHeightMeasureSpec)
    }
}
